src/interface.jl

```diff
@@ -169,8 +193,82 @@ struct Cache{Trule,Ty_cache,Ttangents<:Tuple}
     tangents::Ttangents
 end
 
-_copy!!(dst, src) = copy!(dst, src)
-_copy!!(::Number, src::Number) = src
+"""
+    __exclude_unsupported_output(y)
+
+Required for the robust design of [`value_and_pullback`](@ref), [`prepare_pullback_cache`](@ref).  
+Ensures that `y` contains no aliasing, circular references, `Ptr`s or non differentiable datatypes. 
+In the forward pass f(args...) output can only return a "Tree" like datastructure with leaf nodes as primitive types.  
+Refer https://github.com/compintell/Mooncake.jl/issues/517#issuecomment-2715202789 and related issue for details.  
+Internally calls [`__exclude_unsupported_output_internal!`](@ref).
+The design is modelled after `zero_tangent`.
+"""
+function __exclude_unsupported_output(y::T) where {T}
+    __exclude_unsupported_output_internal!(y, Set{UInt}())
+    return nothing
+end
+
+"""
+    __exclude_unsupported_output_internal(y::T, address_set::Set{UInt}) where {T}
+
+For checking if output`y` is a valid Mutable/immutable composite or a primitive type.
+Performs a recursive depth first search over the function output `y` with an `isbitstype()` check base case. The visited memory addresses are stored inside `address_set`.
+If the set already contains a newly visited address, it errors out indicating an Alais or Circular reference.
+Also errors out if `y` is or contains a Pointer.
+It is called internally by [`__exclude_unsupported_output(y)`](@ref).
+"""
+function __exclude_unsupported_output_internal!(y::T, address_set::Set{UInt}) where {T}
+    isbitstype(T) && return nothing
+    if objectid(y) in address_set
+        throw_circular_reference_or_alias_error(y)
+    end
+
+    # immutable types are copied on the stack.
+    ismutable(y) && push!(address_set, objectid(y))
+
+    # recurse over a composite type's fields.
+    for y_sub in fieldnames(T)
+        # isdefined() is valid for Mutable Structs, Structs.
+        !isdefined(y, y_sub) && continue
+        __exclude_unsupported_output_internal!(getfield(y, y_sub), address_set)
+    end
+
+    return nothing
+end
+
+const IterableCollections = @static VERSION >= v"1.11" ? Union{Array,Memory} : Array
+
+function __exclude_unsupported_output_internal!(
+    y::T, address_set::Set{UInt}
+) where {T<:IterableCollections}
```

Comment:
```suggestion
const __BuiltinArrays = @static VERSION >= v"1.11" ? Union{Array,Memory} : Array

function __exclude_unsupported_output_internal!(
    y::T, address_set::Set{UInt}
) where {T<:__BuiltinArrays}
```
so sorry, would you mind if we renamed `IterableCollections`? It's just a very general name, and I think we would be better off with something more specific. (I otherwise very much like what you've done here to avoid having to duplicate code for `Array`s and `Memory`s)

Reply:
No worries, ive added the change. I agree it does sound too generic, Thanks!